compiler/src/iree/compiler/Codegen/Common/CombineLayoutTransformation.cpp

```diff
@@ -0,0 +1,383 @@
+        mapScatterOp->getLoc(), linearizeIndexOp.getResult(), resultDims);
+    return delinearizeIndexOp->getResults();
+  };
+  rewriter.modifyOpInPlace(mapScatterOp, [&]() {
+    mapScatterOp.insertTransformationAtStart(rewriter, indexTransformBuilder,
+                                             srcDims.size());
+    mapScatterOp->setOperand(0, reshapeOp->getOperand(0));
+  });
+  return mapScatterOp;
+}
+
+static IREE::LinalgExt::MapScatterOp
+foldCollapseShapeIntoMapScatter(RewriterBase &rewriter,
+                                tensor::CollapseShapeOp collapseShapeOp,
+                                IREE::LinalgExt::MapScatterOp mapScatterOp) {
+  assert(mapScatterOp.getInput() == collapseShapeOp->getResult(0) &&
+         "expected collapseShapeOp to be the producer of mapScatterOp");
+  Location loc = collapseShapeOp->getLoc();
+  OpBuilder::InsertionGuard g(rewriter);
+  rewriter.setInsertionPointAfter(collapseShapeOp);
+  SmallVector<OpFoldResult> mixedSrcSizes =
+      tensor::getMixedSizes(rewriter, loc, collapseShapeOp.getSrc());
+  // There can be leftover tensor.dim ops consuming the result of the collpase,
+  // but they will be folded into some affine.apply ops on the source sizes by
+  // later cleanup patterns.
+  SmallVector<OpFoldResult> mixedResultSizes =
+      tensor::getMixedSizes(rewriter, loc, collapseShapeOp.getResult());
+  return foldReshapeIntoMapScatter(rewriter, collapseShapeOp, mixedSrcSizes,
+                                   mixedResultSizes, mapScatterOp);
+}
+
+static IREE::LinalgExt::MapScatterOp
+foldExpandShapeIntoMapScatter(RewriterBase &rewriter,
+                              tensor::ExpandShapeOp expandShapeOp,
+                              IREE::LinalgExt::MapScatterOp mapScatterOp) {
+  assert(mapScatterOp.getInput() == expandShapeOp->getResult(0) &&
+         "expected expandShapeOp to be the producer of mapScatterOp");
+  Location loc = expandShapeOp->getLoc();
+  OpBuilder::InsertionGuard g(rewriter);
+  rewriter.setInsertionPoint(expandShapeOp);
+  SmallVector<OpFoldResult> srcMixedSizes =
+      tensor::getMixedSizes(rewriter, loc, expandShapeOp.getSrc());
+  return foldReshapeIntoMapScatter(rewriter, expandShapeOp, srcMixedSizes,
+                                   expandShapeOp.getMixedOutputShape(),
+                                   mapScatterOp);
+}
+
+/// Fold an `extractSliceOp` into a consumer `mapScatterOp` by applying a mask
+/// based on the bounds of the extractSliceOp. Currently, only zero offsets and
+/// unit strides are supported.
+///
+/// For a given `%mask` value, and slice sizes of `%size0`, `%size1`, ...,
+/// `%sizeN`, the `%new_mask` becomes:
+/// ```
+/// %bound0 = arith.cmpi uge, %idx0, %size0 : index
+/// %mask0 = arith.andi %mask, %bound0 : i1
+/// %bound1 = arith.cmpi uge, %idx1, %size1 : index
+/// %mask1 = arith.andi %mask0, %bound1 : i1
+/// ...
+/// %boundN = arith.cmpi uge, %idxN, %sizeN : index
+/// %new_mask = arith.andi %maskN-1, %boundN : i1
+/// ```
```

Comment:
Optional style nit: I'm not a fan of markdown block style comment, because they don't provide values. They can be blank, we can add few indents for it.

```suggestion
///   %bound0 = arith.cmpi uge, %idx0, %size0 : index
///   %mask0 = arith.andi %mask, %bound0 : i1
///   %bound1 = arith.cmpi uge, %idx1, %size1 : index
///   %mask1 = arith.andi %mask0, %bound1 : i1
///   ...
///   %boundN = arith.cmpi uge, %idxN, %sizeN : index
///   %new_mask = arith.andi %maskN-1, %boundN : i1
```